Librerie/libpdel-0.6.1/pdel/io/filter.3
.\" @COPYRIGHT@
.\"
.\" Author: Archie Cobbs <archie@freebsd.org>
.\"
.\" $Id: filter.3 901 2004-06-02 17:24:39Z archie $
.\"
.Dd April 22, 2002
.Dt FILTER 3
.Os
.Sh NAME
.Nm filter_read ,
.Nm filter_write ,
.Nm filter_end ,
.Nm filter_convert ,
.Nm filter_destroy ,
.Nm filter_fopen ,
.Nm filter_process
.Nd generic data filtering
.Sh LIBRARY
PDEL Library (libpdel, \-lpdel)
.Sh SYNOPSIS
.In sys/types.h
.In stdio.h
.In pdel/io/filter.h
.Ft int
.Fn filter_read "struct filter *f" "void *buf" "int len"
.Ft int
.Fn filter_write "struct filter *f" "const void *data" "int len"
.Ft int
.Fn filter_end "struct filter *f"
.Ft int
.Fn filter_convert "struct filter *f" "int num" "int forward"
.Ft void
.Fn filter_destroy "struct filter **fp"
.Ft "FILE *"
.Fn filter_fopen "struct filter *filter" "int flags" "FILE *fp" "const char *mode"
.Ft "int"
.Fn filter_process "struct filter *filter" "const void *input" "int len" "int final" "u_char **outputp" "const char *mtype"
.Sh DESCRIPTION
These functions operate on
.Em filters ,
which are objects that have an input side and an output side and
perform some kind of encoding or operation on data as it passes through.
.Ss "Filter Objects"
A filter object looks like this:
.Pp
.Bd -literal -offset 3n
struct filter {
    filter_read_t       *read;          /* read data out of filter */
    filter_write_t      *write;         /* write data into filter */
    filter_end_t        *end;           /* signal end of data */
    filter_convert_t    *convert;       /* map # bytes in <-> out */
    filter_destroy_t    *destroy;       /* destroy filter */
    void                *private;       /* object private data */
};
.Pp
.Ed
The
.Va read ,
.Va write ,
.Va end ,
.Va convert ,
and
.Va destroy
fields are pointers to functions having the following types:
.Pp
.Bd -literal -offset 3n
typedef int   filter_read_t(struct filter *f, void *buf, int len);
typedef int   filter_write_t(struct filter *f,
                  const void *data, int len);
typedef int   filter_end_t(struct filter *f);
typedef int   filter_convert_t(struct filter *f,
                  int num, int forward);
typedef void  filter_destroy_t(struct filter **fp);
.Ed
.Pp
.Em Note :
these functions must be implemented to be thread-safe.
For example, two threads should be able to write to and read from
the same filter object simultaneously.
.Pp
The
.Fn read
method should read data from the filter and return the number of bytes
read (up to
.Fa len
bytes),
or 0 if more data needs to be written to the filter (i.e., the filter's
internal buffer is empty).
If an error is encountered, it should return -1 with
.Va errno
set appropriately.
.Pp
The
.Fn write
method inputs data into the filter, returning the number of bytes input
(up to
.Fa len
bytes).
It should return 0 if the filter's internal buffer is full, or -1 and set
.Va errno
on error.
.Pp
The
.Fn end
method indicates to the filter that no more data will be input.
It should return 0 on success or -1 and set
.Va errno
on error.
After this method is called, any calls to
.Fn write
should return -1 with
.Va errno
set to
.Er EPIPE.
.Pp
The
.Fn convert
method provides estimates of the ratio of input length
to output length, and vice-versa.
If
.Fa forward
is non-zero,
.Fn convert
should return an upper bound on the number of bytes of that
.Fa num
bytes of input will generate.
Otherwise, it should return an upper bound on the number of bytes
of input that are required to generate
.Fa num
(or more) bytes of output.
.Pp
The
.Fn destroy
method should free all resources associated with the filter and set
.Fa "*fp"
to
.Dv NULL.
If
.Fa "*fp"
is already equal to
.Dv NULL,
.Fn destroy
should do nothing.
.Pp
The
.Fn filter_read ,
.Fn filter_write ,
.Fn filter_end ,
.Fn filter_convert
and
.Fn filter_destroy
functions are convenience wrappers for the corresponding object methods.
.Pp
By implementing these methods and providing a constructor function
to create new instances, user-defined filters may be used with the
functions below.
.Ss "Filter Functions"
.Fn filter_fopen
pushes a filter on top of a uni-directional stream, returning a new stream.
Data read from or written to the newly created stream will pass
through the filter.
The
.Fa mode
argument is as described for
.Xr fopen 3 ,
but is restricted to being either "r" or "w".
.Pp
If
.Fa flags
is zero, calling
.Xr fclose 3
on the newly created stream causes the underlying stream
.Fa fp
to be closed and
.Fa filter
to be destroyed.
Otherwise, the
.Fa flags
value may contain any of the following values OR'd together:
.Pp
.Bd -literal -offset 3n
FILTER_NO_CLOSE_STREAM    fclose() does not close underlying stream
FILTER_NO_DESTROY_FILTER  fclose() does not destroy the filter
.Ed
.Pp
.Fn filter_process
sends
.Fa len
bytes of data pointed to by
.Fa input
through the filter.
Upon successful return, the results are placed into a newly-allocated
buffer having
.Xr typed_mem 3
type
.Fa mtype
and pointed to by
.Fa "*outputp" ;
the caller must eventually free this buffer.
If
.Fa final
is non-zero, then the filter's
.Fn end
method is called after the last input byte is written.
.Sh RETURN VALUES
All functions that have a return value use -1 or
.Dv NULL
to indicate an error, with
.Va errno
set appropriately.
.Sh SEE ALSO
.Xr base64 3 ,
.Xr fopen 3 ,
.Xr libpdel 3 ,
.Xr string_fp 3 ,
.Xr typed_mem 3
.Sh HISTORY
The PDEL library was developed at Packet Design, LLC.
.Dv "http://www.packetdesign.com/"
.Sh AUTHORS
.An Archie Cobbs Aq archie@freebsd.org


Librerie/libpdel-0.6.1/pdel/io/filter.c

/*
 * @COPYRIGHT@
 *
 * Author: Archie Cobbs <archie@freebsd.org>
 */

#include <sys/types.h>
#include <sys/param.h>

#include <assert.h>
#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <errno.h>
#include <stdlib.h>

#include "structs/structs.h"
#include "structs/type/array.h"

#include "io/filter.h"
#include "util/typed_mem.h"


/************************************************************************
			FILTER STREAMS
************************************************************************/

/* funopen(3) methods used by filter_fopen() */
static int	filter_stream_read(void *cookie, char *buf, int len);
static int	filter_stream_write(void *cookie, const char *buf, int len);
static int	filter_stream_close(void *cookie);

/* State used by filter_fopen() */
struct filter_stream {
	FILE		*fp;			/* underlying stream */
	struct filter	*filter;		/* filter object */
	int		output;			/* input or output? */
	int		eof;			/* eof has been reached */
	int		flags;			/* flags */
};

/*
 * Create a new filter stream using the supplied filter.
 */
FILE *
filter_fopen(struct filter *filter, int flags, FILE *fp, const char *mode)
{
	struct filter_stream *fs;
	int output;

	if (strcmp(mode, "r") == 0)
		output = 0;
	else if (strcmp(mode, "w") == 0)
		output = 1;
	else {
		errno = EINVAL;
		return (NULL);
	}
	if ((fs = MALLOC("filter_stream", sizeof(*fs))) == NULL)
		return (NULL);
	memset(fs, 0, sizeof(*fs));
	fs->filter = filter;
	fs->fp = fp;
	fs->output = output;
	fs->flags = flags;
	if ((fp = funopen(fs,
	    !fs->output ? filter_stream_read : NULL,
	    fs->output ? filter_stream_write : NULL,
	    NULL, filter_stream_close)) == NULL) {
		FREE("filter_stream", fs);
		return (NULL);
	}
	return (fp);
}

static int
filter_stream_read(void *cookie, char *buf, int len)
{
	struct filter_stream *const fs = cookie;
	u_char fbuf[1024];
	int total = 0;
	int flen;
	int num;
	int i;

	/*
	 * Read from underlying stream until we've read "len" bytes
	 * from the filter or the underlying stream returns EOF.
	 */
	for (total = 0; len > 0; ) {

		/* Read any remaining output from the filter */
		if ((num = filter_read(fs->filter, buf, len)) == -1)
			return (-1);
		buf += num;
		len -= num;
		total += num;

		/* No more data to be read from underlying stream? */
		if (fs->eof || len == 0)
			break;

		/*
		 * Read more data from the underlying stream,
		 * but not any more than necessary.
		 */
		flen = filter_convert(fs->filter, len, 0);
		flen = MIN(flen, sizeof(fbuf));
		clearerr(fs->fp);
		if ((num = fread(fbuf, 1, flen, fs->fp)) == 0) {
			if (ferror(fs->fp))
				return (total > 0 ? total : -1);
			fs->eof = 1;
			if (filter_end(fs->filter) == -1)
				return (total > 0 ? total : -1);
		}

		/* Send the data we just read through the filter */
		for (i = 0; i < num; i += flen) {
			if ((flen = filter_write(fs->filter,
			    fbuf + i, num - i)) == -1)
				return (total > 0 ? total : -1);
		}
	}
	return (total);
}

static int
filter_stream_write(void *cookie, const char *buf, int len)
{
	struct filter_stream *const fs = cookie;
	u_char fbuf[1024];
	int flen;

	if ((len = filter_write(fs->filter, buf, len)) == -1)
		return (-1);
	while ((flen = filter_read(fs->filter, fbuf, sizeof(fbuf))) > 0) {
		if (fwrite(fbuf, 1, flen, fs->fp) != flen)
			return (-1);
	}
	return (len);
}

static int
filter_stream_close(void *cookie)
{
	struct filter_stream *const fs = cookie;
	u_char buf[1024];
	int len;

	if (fs->output) {
		(void)filter_end(fs->filter);			/* XXX */
		while ((len = filter_read(fs->filter, buf, sizeof(buf))) > 0) {
			if (fwrite(buf, 1, len, fs->fp) != len)
				break;
		}
		fflush(fs->fp);
	}
	if ((fs->flags & FILTER_NO_CLOSE_STREAM) == 0)
		fclose(fs->fp);
	if ((fs->flags & FILTER_NO_DESTROY_FILTER) == 0)
		filter_destroy(&fs->filter);

	/* Done */
	FREE("filter_stream", fs);
	return (0);
}

/************************************************************************
			FILTER PROCESSING
************************************************************************/

/*
 * Filter data in memory using a filter.
 */
int
filter_process(struct filter *filter, const void *input, int ilen,
	int final, u_char **outputp, const char *mtype)
{
	int nw, w;
	int nr, r;
	int olen;

	/* Allocate buffer big enough to hold filter output */
	olen = filter_convert(filter, ilen, 1) + 10;
	if ((*outputp = MALLOC(mtype, olen)) == NULL)
		return (-1);

	/* Filter data */
	for (w = r = 0; w < ilen; w += nw, r += nr) {
		if ((nw = filter_write(filter,
		    (char *)input + w, MIN(ilen - w, 1024))) == -1)
			goto fail;
		if ((nr = filter_read(filter, *outputp + r, olen - r)) == -1)
			goto fail;
	}
	if (final) {
		if (filter_end(filter) == -1)
			goto fail;
		do {
			if ((nr = filter_read(filter,
			    *outputp + r, olen - r)) == -1)
				goto fail;
			r += nr;
		} while (nr != 0);
	}
	assert(r < olen);

	/* Done */
	(*outputp)[r] = 0;
	return (r);

fail:
	FREE(mtype, *outputp);
	*outputp = NULL;
	return (-1);
}

/************************************************************************
			FILTER METHOD WRAPPERS
************************************************************************/

int
filter_read(struct filter *filter, void *data, int len)
{
	return (*filter->read)(filter, data, len);
}

int
filter_write(struct filter *filter, const void *data, int len)
{
	return (*filter->write)(filter, data, len);
}

int
filter_end(struct filter *filter)
{
	return (*filter->end)(filter);
}

int
filter_convert(struct filter *filter, int num, int forward)
{
	return (*filter->convert)(filter, num, forward);
}

void
filter_destroy(struct filter **filterp)
{
	struct filter *const filter = *filterp;

	if (filter != NULL)
		(*filter->destroy)(filterp);
}

